### PR TITLE
Upload to Azure Storage sample

### LINKED ISSUE #789: Unable to change 4MB photo upload limit
In WebChat, the limit for “Images” is set to 4MB. This is problematic when using phone cameras to upload images for processing.

### PR DESCRIPTION
**Pull request by compulim**
Fixes #789.

### Samples

-  `*`: [Upload to Azure Storage](https://microsoft.github.io/BotFramework-WebChat/20.a.upload-to-azure-storage/), in [#2127](https://github.com/microsoft/BotFramework-WebChat/pull/2127)

## Description

In this demo, we will show you how to customize Web Chat to directly upload file to Azure Storage, and then send the blob URL to the bot for validation and further processing.

## Specific Changes

---

-  [ ] Testing Added

### CODE REVIEW

File: samples/20.a.upload-to-azure-storage/rest-api/public/index.html [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  <head>
+    <meta charset="utf-8" />
+    <link rel="shortcut icon" href="/images/BotServices.png" />
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+    <meta name="theme-color" content="#000000" />
+    <title>Web Chat Upload to Azure Storage Demo</title>
+    <script src="https://cdn.botframework.com/botframework-webchat/4.4.1/webchat.js"></script>
+    <script src="https://unpkg.com/@azure/storage-blob@10.3.0/browser/azure-storage.blob.min.js" integrity="sha384-fsfhtLyVQo3L3Bh73qgQoRR328xEeXnRGdoi53kjo1uectCfAHFfavrBBN2Nkbdf" crossorigin="anonymous"></script>
+
+    <!-- This is for styling Web Chat for demonstration purpose -->
```
Comment (corinagum) [inline samples/20.a.upload-to-azure-storage/rest-api/public/index.html]:
```suggestion
    <!-- This styling is for Web Chat demonstration purposes -->
```

File: samples/20.a.upload-to-azure-storage/rest-api/public/index.html [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        Aborter,
+        AnonymousCredential,
+        BlockBlobURL,
+        StorageURL,
+        uploadBrowserDataToBlockBlob
+      } = window.azblob;
+
+      // Upload a file to Azure Storage by supplying an Object URL and a name.
+      async function uploadFile({ name, url }) {
+        // Today, Web Chat use createObjectURL() to represent the blob.
```
Comment (corinagum) [inline samples/20.a.upload-to-azure-storage/rest-api/public/index.html]:
```suggestion
        // Today, Web Chat uses createObjectURL() to represent the blob.
```

File: samples/20.a.upload-to-azure-storage/rest-api/public/index.html [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+        // Returns the blob URL.
+        return uploadURL;
+      }
+
+      (async function main() {
+        const store = WebChat.createStore({}, ({ dispatch }) => next => action => {
+          const { type } = action;
+
+          // When Web Chat want to send files, we intercept this action and send an event instead.
```
Comment (corinagum) [inline samples/20.a.upload-to-azure-storage/rest-api/public/index.html]:
```suggestion
          // When Web Chat wants to send files, we intercept this action and send an event instead.
```

File: samples/20.a.upload-to-azure-storage/rest-api/public/index.html [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+          if (type === 'WEB_CHAT/SEND_FILES') {
+            (async function () {
+              // Tells the bot that you are uploading files. This is optional.
+              dispatch({ type: 'WEB_CHAT/SEND_TYPING' });
+
+              const files = await Promise.all(action.payload.files.map(({ name, url }) => uploadFile({ name, url })));
+
+              // Today, DirectLineJS will intercept all attachments with URLs from message activity.
+              // Separating the attachment out as a multipart before sending to the bot.
+              // This behavior is to facilitate the createObjectURL() calls, we might change this behavior in later versions of Web Chat.
```
Comment (corinagum) [inline samples/20.a.upload-to-azure-storage/rest-api/public/index.html]:
```suggestion
              // This behavior is to facilitate the createObjectURL() calls. We might change this behavior in later versions of Web Chat.
```

File: samples/20.a.upload-to-azure-storage/rest-api/public/index.html [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+              // In the meanwhile, we workaround this behavior by using an event activity.
```
Comment (corinagum) [inline samples/20.a.upload-to-azure-storage/rest-api/public/index.html]:
```suggestion
              // In the meantime, we workaround this behavior by using an event activity.
```

File: samples/20.a.upload-to-azure-storage/rest-api/public/index.html [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                }
+              });
+            })().catch(err => console.error(err));
+          } else {
+            return next(action);
+          }
+        });
+
+        // This is for obtaining Direct Line token from the bot.
+        // In production system, we should avoid using Direct Line secret to connect to the bot.
```
Comment (corinagum) [inline samples/20.a.upload-to-azure-storage/rest-api/public/index.html]:
```suggestion
        // In a production system, we should avoid using Direct Line secret to connect to the bot.
```

File: samples/20.a.upload-to-azure-storage/rest-api/src/generateDirectLineToken.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+const fetchJSON = require("./utils/fetchJSON");
+
+// Generates a new Direct Line token given the secret.
+module.exports = async function generateDirectLineToken(secret) {
+  // You should consider using Enhanced Direct Line Authentication to protect user ID.
```
Comment (corinagum) [inline samples/20.a.upload-to-azure-storage/rest-api/src/generateDirectLineToken.js]:
```suggestion
  // You should consider using Enhanced Direct Line Authentication to protect the user ID.
```

File: samples/20.a.upload-to-azure-storage/rest-api/src/routes/azureStorage/uploadSASToken.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  const blobName = [
+    now.getUTCFullYear(),
+    pad(now.getUTCMonth() + 1),
+    pad(now.getUTCDate()),
+    pad(now.getUTCHours()),
+    uuidV4()
+  ].join("/");
+  const permissions = new BlobSASPermissions();
+
+  // We only allows create permission, so the user cannot use the URL to redownload the file to redistribute it.
```
Comment (corinagum) [inline samples/20.a.upload-to-azure-storage/rest-api/src/routes/azureStorage/uploadSASToken.js]:
```suggestion
  // We only allow create permissions, so the user cannot use the URL to redownload the file to redistribute it.
```

File: samples/20.a.upload-to-azure-storage/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# This container is for simplifying CI when using Azure Pipelines
+
+# Aggregates all code into a single Docker image for export
+FROM node:12
+
+# Copy the bot code to /var/bot/
+ADD bot/ /var/build/bot/
+
+# Copy the web server code to /var/web/
+ADD rest-api/ /var/build/web/
```
[the remark this answers is not in the document]
Comment (compulim) [reply samples/20.a.upload-to-azure-storage/Dockerfile]: Renamed from `rest-api` to `web`.

File: samples/20.a.upload-to-azure-storage/bot/package.json [hunk 10]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "single-sign-on-bot",
```
[the remark this answers is not in the document]
Comment (compulim) [reply samples/20.a.upload-to-azure-storage/bot/package.json]: Renamed from `single-sign-on-bot` to `bot`.

**Comment by compulim** [comment]
@corinagum I have updated the PR with typos, renamed `rest-api` to `web`, and renamed `single-sign-on-bot` to `bot`.

Please review.

File: samples/20.a.upload-to-azure-storage/README.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+You can follow our instructions on how to [setup a new Bot Channel Registration](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-quickstart-registration?view=azure-bot-service-3.0).
+
+1. Save the Microsoft App ID and password to `/bot/.env`
+   -  `MICROSOFT_APP_ID=12345678-1234-5678-abcd-12345678abcd`
+   -  `MICROSOFT_APP_PASSWORD=a1b2c3d4e5f6`
+1. Save the Web Chat secret to `/web/.env`
+   -  `DIRECT_LINE_SECRET=a1b2c3.d4e5f6g7h8i9j0`
+
+> When you are building your production bot, never expose your Web Chat or Direct Line secret to the client. Instead, you should use the secret to generate a limited token and send it to the client. For information, please refer [to this page on how to generate a Direct Line token](https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-direct-line-3-0-authentication?view=azure-bot-service-4.0#generate-token) and [Enhanced Direct Line Authentication feature](https://blog.botframework.com/2018/09/25/enhanced-direct-line-authentication-features/).
```
Comment (corinagum) [inline samples/20.a.upload-to-azure-storage/README.md]:
```suggestion
> When you are building your production bot, never expose your Web Chat or Direct Line secret to the client. Instead, you should use the secret to generate a limited token and send it to the client. For information, please refer the documentation on [Direct Line token generation](https://docs.microsoft.com/en-us/azure/bot-service/rest-api/bot-framework-rest-direct-line-3-0-authentication?view=azure-bot-service-4.0#generate-token) and the [Enhanced Direct Line Authentication feature](https://blog.botframework.com/2018/09/25/enhanced-direct-line-authentication-features/).
```

**Comment by akbloodadarsh** [comment]
@compulim Hello, do you have a c# version of the "upload to azure storage file"

**Pull request merged**